Fixed bad base path edge case

loren-osborn opened the pull request:
This is a fix to the issue I brought up in #3433 . 

While the code is a bit rough around the edges, it appears to conform to project guidelines, and resolves the issue. Any suggestions to make the code more elegant are welcome.

loren-osborn commented on src/Codeception/Configuration.php:
```diff
@@ -500,6 +500,58 @@ public static function projectDir()
+                $relPath = substr($path, $rootPathIndex);
+                $pathPrefix = substr($pathPrefix, $rootPathIndex);
+                $basePathArr = explode(DIRECTORY_SEPARATOR, $pathPrefix);
+                while (
```
... because of overly complex while condition. Please make code simplification suggestions. (@Nitpick-CI, Not you personally, I know you're just an automated tool.)

DavertMik commented on src/Codeception/Configuration.php:
```diff
@@ -500,6 +500,58 @@ public static function projectDir()
+    {
+        $relPath = $path;
+        if (substr($path, 0, strlen(self::projectDir())) == self::projectDir()) {
+            $relPath = substr($path, strlen(self::projectDir()));
```
don't do `else`, just return it here if you already matched it

DavertMik commented on src/Codeception/Configuration.php:
```diff
@@ -500,6 +500,58 @@ public static function projectDir()
+            $sameAbsoluteness = true;
+            $rootPathIndex = 0;
+            if (DIRECTORY_SEPARATOR == '\\') {
+                $hasDriveSpecifier = preg_match('/^[A-Za-z]:/', $pathPrefix);
```
please add some comments here on why regex is using and what is matched

loren-osborn replied:
I know there are a couple regexes... This one matches Windows style drive letters.

DavertMik commented on src/Codeception/Configuration.php:
```diff
@@ -500,6 +500,58 @@ public static function projectDir()
+                    $sameAbsoluteness = ($pathPrefix[0] == $path[0]);
+                }
+            }
+            if ($sameAbsoluteness) {
```
add some comments here too... I can't understand the logic :|

loren-osborn replied:
I mentioned when I submitted the PR, that the code was pretty rough. Basically you can only make a relative path of one path against the other if they both have the same "absoluteness":

* On *nix this is easy:
    * A path is absolute if it starts with a slash (a Boolean)
    * Two paths have the same absoluteness if both or neither start with a slash
* On Windows this is less clear:
    * a path can be relative to the current or a specified device
    * it can also be absolute or relative to the CWD of that device
    * If a drive is specified, they both have to specify the same drive
    * If no drive is specified, they both need to *not* specify a drive
    * whether or not they specify a drive letter the first non-drive letter character (after the `:`) need to both or neither start with a backslash.

If the paths don't have the same absoluteness, you can't make a relative path because they're *not* relative to one another... 

I need to clarify the code to capture that more clearly, but does that clarify the logic to you at least?

DavertMik commented on src/Codeception/Configuration.php:
```diff
@@ -500,6 +500,58 @@ public static function projectDir()
+                     ($path[$rootPathIndex]       == DIRECTORY_SEPARATOR));
+                $rootPathIndex += ($pathPrefix[$rootPathIndex] == DIRECTORY_SEPARATOR) ? 1 : 0;
+            }
+            if ($sameAbsoluteness) {
```
comments! what are you trying to achieve here too

loren-osborn replied:
I will add comments:

In advance of that, I'm peeling off all common parts of the path from left of both paths, and for all unmatched directory remaining pretending a `..` directory to the result.

DavertMik commented on src/Codeception/Configuration.php:
```diff
@@ -500,6 +500,58 @@ public static function projectDir()
+                        $relPath = '..' . DIRECTORY_SEPARATOR . $relPath;
+                    }
+                }
+            } else {
```
don't do `else`

loren-osborn replied:
Instead end the if and the function with distinct `return`s?

loren-osborn commented:
Thanks for the feedback... I will revise the code as soon as I'm able.

DavertMik commented:
Thanks, will be waiting for update!

loren-osborn commented:
@DavertMik, Thanks... I will update this PR as soon as I can find time to update it. I didn't see a way to PM you, but wanted to know if you could take a peek at an issue I think I found: https://github.com/Codeception/Codeception/compare/2.2...loren-osborn:losborn_configModsLooksLikeABug  It shouldn't be hard to write a unit test to catch this issue... (If a enabled module in the module `.yml` file has options, the Codeception should [incorrectly] try to enable a module with the name of the module's first option, instead of its own name.)

loren-osborn commented:
@DavertMik, let me know if you found this much easier to follow. I prefer descriptive variable and method names to comments, but I included good comments too.

DavertMik commented:
@loren-osborn looks etter now, however I still have doubts. I think code still can be simplified but I'm not sure how :no_mouth:

Could you add tests to it? They should go to `tests/unit/ConfigurationTest.php`. 

@Naktibalda @sergeyklay what your thoughts of it?

sergeyklay commented:
+1 tests

loren-osborn commented:
I'm certainly down with tests... It's more a matter of knowing what kind of tests, and where to find them. I just found the right place for unit tests for this particular issue, and the issue I found with `Codeception\Configuration::modules()`:

* This file: `Codeception/tests/unit/Codeception/ConfigurationTest.php` for unit tests.

Where else do you suggest I put tests for these two issues. I currently don't have a setup where I can run the tests (might I suggest a `Vagrantfile` with a Puppet, Chef or Ansible configuration) so I'll have to rely on Travis and the other CI services.

Naktibalda commented:
I found `getRelativeDir()` very hard to read.

I think that it would be much easier to unit test it, if project dir was passed in as a parameter:
`public static function getRelativeDir($path, $projectDir)`

```php
function codecept_relative_path($path)
{
    return \Codeception\Configuration::getRelativeDir($path, \Codeception\Configuration::projectDir());
}
```

All you need to run ConfigurationTest is PHP,
but if you want to run all tests, try to use Docker as documented in https://github.com/Codeception/Codeception/blob/2.2/tests/README.md

loren-osborn commented:
@Naktibalda, passing the project directory as an argument is a great idea. I was trying to keep the `codecept_relative_path()` API, but as you pointed out there is no need for that.

As far as a dev environment, I currently have limited storage to keep multiple VMs around, and my current VM is missing a PECL module that composer requires for Codeception. (I don't want to alter one project's dev environment VM for development of another.) I'll see what I can put together if I have the time, but I won't make any promises with my looming deadlines.

loren-osborn commented:
I think I took what @Naktibalda said to heart. I rewrote this so I can inject both `projectDir()` *and* `DIRECTORY_SEPARATOR` so I will now be able to test the Windows path handling without being on a Windown system. I also rewrote the `$path` and `$projDir`handling for clairity, so both are either being manipulated as strings *or* as arrays, not one of each. I still need to write the actual tests, but that should be straightforward now.

I'm at dinner now, but hope to

loren-osborn commented:
Status update: Wrote some tests, found some edge casses I missed. I'll address before I push my code.

loren-osborn commented:
Enabled test cases... disabled code changes... test suite *should FAIL*

loren-osborn commented:
@DavertMik, @sergeyklay, @Naktibalda, I think the `semaphoreci` build needs to be restarted...

Once I get a signoff from one of you, I can get started on a git rebase squash of this code. (Of course, I'm still open to suggestions. It's far from the most readable thing I've written.)

loren-osborn commented:
All changed squashed into single commit.  Original unsquashed commits are still available here: https://github.com/loren-osborn/Codeception/tree/losborn_badBasePathFix_unsquashed

Thanks for all of your help. Do any of you want any additional revisions before merging this? (i.e. It may be a good idea to build the output array returned by `ConfigurationTest::getRelativeDirTestData()` to be generated as a permutation of several test conditions, but I'm unsure if that would be more or less readable. Any feedback is appreciated.)

Naktibalda commented on src/Codeception/Configuration.php:
```diff
@@ -500,6 +500,126 @@ public static function projectDir()
+    public static function getRelativeDir($path, $projDir, $dirSep = DIRECTORY_SEPARATOR)
+    {
+        // ensure $projDir ends with a trailing $dirSep
+        $projDir = preg_replace('/'.preg_quote($dirSep, '/').'*$/', $dirSep, $projDir);
```
No an issue, but mu preferred way is $projDir = rtrim($projDir, $dirSep) . $dirSep;

loren-osborn replied:
Yes... Probably clearer your way.

DavertMik replied:
Yeah, don't use regex here. rtrim is much clearer

Naktibalda commented on src/Codeception/Configuration.php:
```diff
@@ -500,6 +500,126 @@ public static function projectDir()
+        }
+        // peel off optional absoluteness prefixes and convert
+        // $path and $projDir to an subdirectory path array
+        $relPathParts = array_filter(explode($dirSep, substr($path, strlen($pathAbsPrefix['wholePrefix']))), 'strlen');
```
'strlen' is not necessary, array_filter removes empty strings when callback is not provided.

loren-osborn replied:
Agree... Was trying to make the code more clear... I probably failed. ;-)

loren-osborn replied:
I take back my previous comment... Without a callback, array_filter is supposed to remove any element that evaluates to FALSE... I believe this includes some non-empty strings, including `"0"`... Looks like I should add some dies named `0` to my data provider test cases.

Naktibalda replied:
then using strlen is fine.

Naktibalda commented on tests/unit/Codeception/ConfigurationTest.php:
```diff
@@ -67,4 +67,83 @@ public function testDefaultCustomCommandConfig()
+                    ['\\my\\proj\\path\\some\\dir\\in\\my\\proj\\',      '\\my\\proj\\path\\foobar\\',   '\\',  '..\\some\\dir\\in\\my\\proj\\'],
+                // Device letter (both)... path absoluteness mismatch
+                    ['C:\\my\\proj\\path\\some\\file\\in\\my\\proj.txt', 'C:my\\proj\\path\\',           '\\',  '\\my\\proj\\path\\some\\file\\in\\my\\proj.txt'],
+                    ['d:my\\proj\\path\\some\\file\\in\\my\\proj.txt',   'd:\\my\\proj\\path\\',         '\\',  'my\\proj\\path\\some\\file\\in\\my\\proj.txt'],
```
Is `d:my` a valid path on Windows?

loren-osborn replied:
To my knowledge it is... It certainly was in the MSDOS days... Each drive letter keeps its own CWD

Naktibalda commented on tests/unit/Codeception/ConfigurationTest.php:
```diff
@@ -67,4 +67,83 @@ public function testDefaultCustomCommandConfig()
+        return [
+            // Unix style paths:
+                // projectDir() with & without trailing directory seperator: actual subdir
+                    ['/my/proj/path/some/file/in/my/proj.txt',           '/my/proj/path/',               '/',  'some/file/in/my/proj.txt'],
```
The size of this data provider is impressive, so it must be good.

loren-osborn replied:
I believe you were being sarcastic, but I disagree (so actually agreeing)... I think shorter is clearer... Just not sure how to pair this down better

loren-osborn replied:
... Each device has its own CWD on a per-shell basis, I believe

DavertMik replied:
Yes, the list is impressive. And that's really good

loren-osborn commented:
@DavertMik, @sergeyklay, @Naktibalda, Status update:

* I have confirmed that `D:Foo` is a perfectly valid Windows path (referring to file or directory `Foo` in this shell's current working directory for device `D:`)
* While I'm open to some tweaking, I think this patch is basically done and ready to merge.
* Welcome additions include:
    * Additional test cases with `0` as one of the directories in paths
    * Easier to read generation of test cases instead of an confusing exhaustive list
* I can get to these when I have time, but consider them low priority.

Naktibalda commented:
Your PR looks good to me, I am waiting for somebody else to review the code before merging it.

loren-osborn commented:
Thanks

DavertMik commented:
Thanks. I think you did a great job and this PR looks almost perfect to me. Especially tests look great.
However, now I have a feeling that `isWindows`, `getPathAbsolutenessPrefix`, and other added methods may not belong to Configuration class. They are not about Configuration, right? That's why I'd suggest to move the path resolving logic to a utility class, probably `Codeception\Util\PathResolver` and make Configuration only call them to get relative path.

Sorry for this additional request but that's how we can maintain the clean and sane codebase.

loren-osborn commented:
This seems like a very reasonable request , though I'm leaning toward `Codeception\Util\PathHandler` unless the other already exists. I don't expect address this this week, but will try to get to it shortly. @DavertMik, if you can suggest a more readable way to generate the given test cases, I would be grateful.

DavertMik commented:
Looks good. Please proceed

Naktibalda commented:
@loren-osborn you are about to miss a release.

DavertMik commented:
Thanks @loren-osborn 
We will extract class by ourselves and your fix is about to land in next release

loren-osborn commented:
@DavertMik, @Naktibalda thanks for picking up the slack on this. My last comment in this PR was actually in the few days between when I was laid off from my last employer and when my daughter was born, so you can imagine I was a bit distracted. I ended up being out of work about 6 months and didn't realize you had actually merged this. Thank you for picking up the slack in my absence.
